src/Neo/VM/Helper.cs

```diff
@@ -364,13 +364,14 @@ private static JObject ToJson(StackItem item, HashSet<StackItem> context, ref in
             {
                 case Array array:
                     {
-                        context ??= new HashSet<StackItem>(ReferenceEqualityComparer.Instance);
+                        context ??= new(ReferenceEqualityComparer.Instance);
                         if (!context.Add(array)) throw new InvalidOperationException("Circular reference.");
                         maxSize -= 2/*[]*/+ Math.Max(0, (array.Count - 1))/*,*/;
                         JArray a = new();
                         foreach (StackItem stackItem in array)
                             a.Add(ToJson(stackItem, context, ref maxSize));
                         value = a;
+                        if (!context.Remove(array)) throw new InvalidOperationException("Circular reference.");
```

Comment:
This case is a bit different from SC parameter, because I can construct such VM item with circular reference with the help of VM opcodes, and then I can use native StdLib's `jsonSerialize` method to serialize this item.

So this VM-related fix requires a hardfork. I checked NeoGo behaviour, and we don't have such bug but it's just because our stackitem serializator does not *explicitly* track circular references (we have size constraint tracker instead). And since Go/C# nodes state currently match in both testnet/mainnet, it means that this code was not yet triggered on these networks. But still, it may be triggered by transaction, hence it should be moved under Echidna.

Reply:
OK, I just checked C# implementation of StdLib's `jsonSerialize` and turns out it uses different serialization method:
https://github.com/neo-project/neo/blob/7f5504090420baf18175a485e959f268b0d11712/src/Neo/SmartContract/Native/StdLib.cs#L47

>  it seems that is not called inside the VM

So you're right, no fork required.